### PR TITLE
Feature/106_development_filters

### PR DESCRIPTION
**Pull request by coxen34**
**Description**

A new optional tag parameter has been added.
This parameter provides users the ability to further refine their queries and obtain more specific results based on the tags associated with the students.

The documentation has been updated and a test has been added.

**Refactorization:** Modularization of the creation of 'student' & 'resume' in tests

- Created `Tests\Fixtures\Students` `Tests\Fixtures\Resumes`  classes to encapsulate the logic of creating 'students' & 'resumes' in the tests.
- Static method additions
  `aStudent()` `createResume()` to generate 'student' and 'resumes' in a cleaner and more reusable manner.
-Now utilized:
   `'Students::aStudent()'` `Resumes::createResume()` instead of instantiating directly in testing, which improves the 
     readability and maintainability of the code.
- This refactoring makes it easier to create different types of 'resumes' by passing the `specialization` parameter.

This simplifies testing and promotes a better modular structure of the code, which will make future updates and maintenance easier.

### CODE REVIEW

**Review by toniGitH — COMMENTED** [review]
I think the code works correctly. :+1: Great!!

However, I would recommend two changes before approving it:

1) It would be convenient to **remove the 'specialization' field from the JSON response** (which was probably added to facilitate testing different cases in Swagger and was left without being deleted).

2) The **Swagger information about the URL needed to filter by specializations+tags** (both at the same time) should contain the names of the tags and not their IDs, since the endpoint works with the names, but not with the IDs.
So, where it says
```To filter by both specialization and tags: /student/list/for-home?specialization=frontend&tags=1,2```
it should say
```To filter by both specialization and tags: /student/list/for-home?specialization=frontend&tags=PHP,Laravel```.

File: app/Service/StudentListService.php [hunk 1]
```diff
@@ -12,21 +12,33 @@
...
+        $query = Resume::query();
 
-        if ($specializations[0] != null) {
-            $resumes = Resume::whereIn('specialization', $specializations)->get();
-        } else {
-            $resumes = Resume::all();
+        if ($specializations !== null && $specializations[0] != null) {
+            $query->whereIn('specialization', $specializations);
         }
+        if ($tags != null) {
```
Comment (jordimorillo) [inline app/Service/StudentListService.php:29]: !==

File: tests/Feature/Student/StudentListTest.php [hunk 2]
```diff
@@ -83,4 +84,36 @@ public function test_specialization_filter_returns_correct_students()
         $this->assertNotNull($fetchedResume);
         $this->assertTrue($fetchedResume->student->is($student));
     }
+
+    public function testGetResumesWithTags()
+    {
+        $tag1 = Tag::create(['tag_name' => 'tag1']);
+        $tag2 = Tag::create(['tag_name' => 'tag2']);
+        $student = Student::factory()->create([
```
Comment (jordimorillo) [inline tests/Feature/Student/StudentListTest.php:92]: Per simplificar la lectura, pots extreure el métode amb el que instancies el `$student` a una classe `Tests\Fixtures\Students` que tingui un métode estàtic que instancii el estudiant:

```
# \Tests\Fixtures\Students.php
<?php

declare(strict_types=1);

namespace Tests\Fixtures;

class Students {
    public static function aStudent(): Student
    {
        return Student::factory()->create([
            'name' => 'Dokuta',
            'surname' => 'Suranpu',
            'photo' => 'Arale',
            ]);
    }
}
```

Un cop que tens aquesta clase, per instanciar-ho, només has de fer:

```
$student = Students::aStudent();
```

Et quedarà més net i tindràs un recurs que podràs cridar sempre que necessitis tenir un Student a qualsevol test.

File: tests/Feature/Student/StudentListTest.php [hunk 3]
```diff
@@ -83,4 +84,36 @@ public function test_specialization_filter_returns_correct_students()
...
+            'name' => 'Dokuta',
+            'surname' => 'Suranpu',
+            'photo' => 'Arale'
+        ]);
+        $resume1 = Resume::factory()->create([
```
Comment (jordimorillo) [inline tests/Feature/Student/StudentListTest.php:97]: El mateix amb els resumes. En aquest cas pots veure quina variable et canvia i passar-la com a paràmetre de la funció estàtica, o crear un métode per cada tipus de Resume, per exemple `Resumes::aFrontendResume`, `Resumes::aBackendResume`. Serà més fàcil entendre qué estàs instanciant.

**Review by jordimorillo — APPROVED** [review]
El teu PR està ben fet, Rosa, però t'he escrit alguns consells per si vols fer que el teu codi quedi una mica més net. Bona feina! :)

**Comment by coxen34** [comment]
> I think the code works correctly. 👍 Great!!
> 
> However, I would recommend two changes before approving it:
> 
> 1. It would be convenient to **remove the 'specialization' field from the JSON response** (which was probably added to facilitate testing different cases in Swagger and was left without being deleted).
> 2. The **Swagger information about the URL needed to filter by specializations+tags** (both at the same time) should contain the names of the tags and not their IDs, since the endpoint works with the names, but not with the IDs.
>    So, where it says
>    `To filter by both specialization and tags: /student/list/for-home?specialization=frontend&tags=1,2`
>    it should say
>    `To filter by both specialization and tags: /student/list/for-home?specialization=frontend&tags=PHP,Laravel`.

Thank you Antonio!
Fixed up!!

**Comment by coxen34** [comment]
> El teu PR està ben fet, Rosa, però t'he escrit alguns consells per si vols fer que el teu codi quedi una mica més net. Bona feina! :)

Gràcies Jordi, faré els canvis!

**Review by toniGitH — APPROVED** [review]
Agree with everything. Very interesting the use of fixtures for tests. We'll have to keep that in mind for the future.

**Review by FranEnLaNube — COMMENTED** [review]
Good job @coxen34! I'd love how did you use Markdown in `../controllersAnnotations/studentAnnotations/AnnotationsStudentList.php`.
I couldn't understand enough the real utility of `tests/Fixtures/Students.php`, `tests/Fixtures/Resumes.php` and `aStudent()` method. Can't that be made with another existing files? 
Anyway, this is merged and that just gonna be to learn more for future jobs. 💪 

---

Update: Now I see that it was a @jordimorillo. That doesn't mean that I understand it but, is something that I'll need to check deeper.

**Pull request merged**